src/plots/cartesian/axes.js

```diff
@@ -1636,6 +1636,24 @@ axes.draw = function(gd, arg, opts) {
  * @param {object} ax (full) axis object
  * @param {object} opts
  * - @param {boolean} skipTitle (set to true to skip axis title draw call)
+ *
+ * Depends on:
+ * - ax._mainSubplot (from linkSubplots)
+ * - ax._mainAxis
+ * - ax._anchorAxis
+ * - ax._subplotsWith
+ * - ax._counterDomainMin, ax._counterDomainMax (optionally, from linkSubplots)
+ * - ax._mainLinePosition (from lsInner)
+ * - ax._mainMirrorPosition
+ * - ax._linepositions
```

Comment:
Computing and stashing `_mainLinePosition`, `_mainMirrorPosition` and `_linepositions` in `lsInner` is less than ideal. This makes `Axes.drawOne` "depend" on `lsInner` i.e. `lsInner` must get called before `Axes.drawOne` on first draw and `'ticks'` edits.

---

Now in relation to https://github.com/plotly/plotly.js/issues/2704, if we choose to add a "pure margin computation" step e.g. called `Axes.pushMargin`, we'll need to un-coupled `drawOne` and `lsInner`. After this PR, `Axes.pushMargin` for a single axis could look like:

```js
axes.pushMargin = function(gd, ax) {
  var seq = [];
  
  /* compute mainLinePosition */

  seq.push(function() {
    return drawLabels(gd, ax, {
      vals: axes.calcTicks(ax),
      layer: /* some dummy layer in the tester div */,
      transFn: axes.makeTransFn(ax),
      labelFns: axes.makeLabelFns(ax, mainLinePosition)
    })
  })

  /* N.B. we don't need to draw ticks, grid lines nor axis lines
     to compute the margin push values */

  seq.push(function() {
    /* compute the margin push values */ 

    Plots.autoMargin(gd, /* */) 
  })
}
```

or maybe we could have a set of getter `Axes` helpers so we won't have to stash the values.